diff --git a/src/main/java/com/sap/oss/phosphor/fosstars/advice/AdviceContentYamlStorage.java b/src/main/java/com/sap/oss/phosphor/fosstars/advice/AdviceContentYamlStorage.java
index 0e9c1ed6..4c82ae6e 100644
--- a/src/main/java/com/sap/oss/phosphor/fosstars/advice/AdviceContentYamlStorage.java
+++ b/src/main/java/com/sap/oss/phosphor/fosstars/advice/AdviceContentYamlStorage.java
@@ -1,6 +1,7 @@
 package com.sap.oss.phosphor.fosstars.advice;
 
 import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonGetter;
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.fasterxml.jackson.core.type.TypeReference;
 import com.sap.oss.phosphor.fosstars.model.Feature;
@@ -152,6 +153,23 @@ public class AdviceContentYamlStorage {
       this.name = name;
       this.url = url;
     }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o.getClass() != getClass()) {
+        return false;
+      }
+      RawLink rawLink = (RawLink) o;
+      return Objects.equals(name, rawLink.name) && Objects.equals(url, rawLink.url);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(name, url);
+    }
   }
 
   /**
@@ -183,10 +201,10 @@ public class AdviceContentYamlStorage {
      */
     @JsonCreator
     RawAdviceContent(
-        @JsonProperty("advice") String advice, @JsonProperty("links") List<RawLink> links) {
+        @JsonProperty("advice") String advice,
+        @JsonProperty(value = "links") List<RawLink> links) {
 
       Objects.requireNonNull(advice, "Oh no! Advice is null!");
-      Objects.requireNonNull(advice, "Oh no! Links is null!");
 
       advice = advice.trim();
       if (advice.isEmpty()) {
@@ -194,7 +212,17 @@ public class AdviceContentYamlStorage {
       }
 
       this.advice = advice;
-      this.links = new ArrayList<>(links);
+      this.links = new ArrayList<>(links != null ? links : Collections.emptyList());
+    }
+
+    @JsonGetter("advice")
+    private String advice() {
+      return advice;
+    }
+
+    @JsonGetter("links")
+    private List<RawLink> links() {
+      return new ArrayList<>(links);
     }
 
     /**
@@ -273,5 +301,22 @@ public class AdviceContentYamlStorage {
       return string;
     }
 
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o.getClass() != getClass()) {
+        return false;
+      }
+      RawAdviceContent content = (RawAdviceContent) o;
+      return Objects.equals(advice, content.advice)
+          && Objects.equals(links, content.links);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(advice, links);
+    }
   }
 }
diff --git a/src/test/java/com/sap/oss/phosphor/fosstars/advice/AdviceContentYamlStorageTest.java b/src/test/java/com/sap/oss/phosphor/fosstars/advice/AdviceContentYamlStorageTest.java
index 1ebbe273..9a218b6c 100644
--- a/src/test/java/com/sap/oss/phosphor/fosstars/advice/AdviceContentYamlStorageTest.java
+++ b/src/test/java/com/sap/oss/phosphor/fosstars/advice/AdviceContentYamlStorageTest.java
@@ -4,6 +4,7 @@ import static com.sap.oss.phosphor.fosstars.advice.oss.OssAdviceContentYamlStora
 import static com.sap.oss.phosphor.fosstars.model.feature.example.ExampleFeatures.NUMBER_OF_COMMITS_LAST_MONTH_EXAMPLE;
 import static com.sap.oss.phosphor.fosstars.model.feature.example.ExampleFeatures.SECURITY_REVIEW_DONE_EXAMPLE;
 import static com.sap.oss.phosphor.fosstars.model.feature.example.ExampleFeatures.STATIC_CODE_ANALYSIS_DONE_EXAMPLE;
+import static java.util.Arrays.asList;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
@@ -13,8 +14,8 @@ import com.sap.oss.phosphor.fosstars.advice.AdviceContentYamlStorage.RawLink;
 import com.sap.oss.phosphor.fosstars.advice.oss.OssAdviceContentYamlStorage.OssAdviceContext;
 import com.sap.oss.phosphor.fosstars.model.RatingRepository;
 import com.sap.oss.phosphor.fosstars.model.rating.example.SecurityRatingExample;
+import com.sap.oss.phosphor.fosstars.util.Yaml;
 import java.io.IOException;
-import java.util.Arrays;
 import java.util.List;
 import java.util.Set;
 import org.junit.Test;
@@ -37,9 +38,13 @@ public class AdviceContentYamlStorageTest {
     }
 
     advice = storage.adviceFor(STATIC_CODE_ANALYSIS_DONE_EXAMPLE, EMPTY_OSS_CONTEXT);
-    assertEquals(1, advice.size());
+    assertEquals(3, advice.size());
     assertFalse(advice.get(0).text().isEmpty());
     assertFalse(advice.get(0).links().isEmpty());
+    assertFalse(advice.get(1).text().isEmpty());
+    assertTrue(advice.get(1).links().isEmpty());
+    assertFalse(advice.get(2).text().isEmpty());
+    assertTrue(advice.get(2).links().isEmpty());
 
     advice = storage.adviceFor(NUMBER_OF_COMMITS_LAST_MONTH_EXAMPLE, EMPTY_OSS_CONTEXT);
     assertTrue(advice.isEmpty());
@@ -49,7 +54,7 @@ public class AdviceContentYamlStorageTest {
   public void testRawContentVariables() {
     RawAdviceContent content = new RawAdviceContent(
         "Test ${FIRST_VARIABLE} variable",
-        Arrays.asList(
+        asList(
             new RawLink("First link ${SECOND_VARIABLE}}", "https://test"),
             new RawLink("Second link", "${THIRD_VARIABLE}}")));
 
@@ -59,4 +64,19 @@ public class AdviceContentYamlStorageTest {
     assertTrue(variables.contains("SECOND_VARIABLE"));
     assertTrue(variables.contains("THIRD_VARIABLE"));
   }
+
+  @Test
+  public void testYamlSerialization() throws IOException {
+    RawAdviceContent fullContent = new RawAdviceContent(
+        "Test",
+        asList(
+            new RawLink("First", "https://test/first"),
+            new RawLink("Second", "https://test/second")));
+    RawAdviceContent clone = Yaml.read(Yaml.toBytes(fullContent), RawAdviceContent.class);
+    assertEquals(fullContent, clone);
+
+    RawAdviceContent noLinks = new RawAdviceContent("No links", null);
+    clone = Yaml.read(Yaml.toBytes(noLinks), RawAdviceContent.class);
+    assertEquals(noLinks, clone);
+  }
 }
\ No newline at end of file
diff --git a/src/test/resources/com/sap/oss/phosphor/fosstars/advice/AdviceContentStorageTest.yml b/src/test/resources/com/sap/oss/phosphor/fosstars/advice/AdviceContentStorageTest.yml
index 513c07f0..d01ce4d5 100644
--- a/src/test/resources/com/sap/oss/phosphor/fosstars/advice/AdviceContentStorageTest.yml
+++ b/src/test/resources/com/sap/oss/phosphor/fosstars/advice/AdviceContentStorageTest.yml
@@ -19,3 +19,6 @@ Static code analysis status (example):
         url: https://test/5
       - name: link6
         url: https://test/6
+  - advice: Advice with empty links
+    links: []
+  - advice: Advice with no links
\ No newline at end of file
